app/furniture/marketplace/marketplaces/_marketplace.html.erb

```diff
@@ -1,5 +1,13 @@
 
-<%- cart = marketplace.carts.find_or_create_by(id: session[:current_cart] ||= SecureRandom.uuid) %>
+
+
+<%- shopper = if current_person.is_a?(Guest) 
+     Marketplace::Shopper.find_or_create_by(id: session[:current_cart] ||= SecureRandom.uuid)
+  else
+    Marketplace::Shopper.find_or_create_by(person: current_person)
+end %>
```

Comment:
This is fine for now, but it feels like this might want to eventually live in a controller helper method, instead of the template (and maybe also the cart?). It just seems like a fair amount of non-trivial logic to have in a view.

Reply:
I tried moving this to a helper in `app/furniture/marketplace/marketplaces_controller.rb`  but it forced me to repeat the partial `_marketplace.html.erb` which is used in at least 3 controllers. This would lead to a lot of duplicate code. 

@anaulin @zspencer  Should I just move it to application helper like below or leave the code as in `_marketplace.html.erb`.

```
module ApplicationHelper
  def shopper
    params = if current_person.is_a?(Guest) 
      {id: session[:current_cart] ||= SecureRandom.uuid}
    else
      {person: current_person}
    end

    Marketplace::Shopper.find_or_create_by(params)
  end
end

```

Reply:
Putting it in the `ApplicationHelper` would take over the name `shopper` in all execution contexts; which seems less than ideal.

WHAT IF!

We [generated a `MarketplaceComponent`](https://viewcomponent.org/guide/generators.html) that takes an instance of a `Marketplace::Marketplace` and exposes a `shopper` method? It could access `current_person` and `session` through the [Helpers Proxy](https://viewcomponent.org/guide/helpers.html#proxy)